[amazonechocontrol] last spoken text channel

Trinitus01 opened the pull request:
changed: last voice command via customer history record
added: last spoken text channel

Signed-off-by: Tom Blum <trinitus01@googlemail.com>

Trinitus01 committed:
changed: last voice command via customer history record
added: last spoken text channel

Signed-off-by: Tom Blum <trinitus01@googlemail.com>

J-N-K commented on bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/AmazonEchoControlBindingConstants.java:
```diff
@@ -120,82 +121,44 @@
 
     // DeviceTypeIds to human readable description
     // originally found here: https://github.com/Apollon77/ioBroker.alexa2/blob/master/main.js
-    public static final Map<String, String> DEVICE_TYPES = Map.<String, String> ofEntries( //
```
Please revert this change.

J-N-K commented on bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/Connection.java:
```diff
@@ -1150,6 +1153,21 @@ public String getDeviceListJson() throws IOException, URISyntaxException, Interr
+            String json = makeRequestAndReturnString(
+                    "https://" + amazonSite + "/alexa-privacy/apd/rvh/customer-history-records?startTime="
+                            + (startTime != null ? startTime : "") + "&endTime=" + (endTime != null ? endTime : "")
+                            // + "&recordType=VOICE_HISTORY"
```
remove commented code

J-N-K commented on bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java:
```diff
@@ -785,12 +786,35 @@ private void handlePushActivity(@Nullable String payload) {
-                        .map(sourceDeviceId -> findEchoHandlerBySerialNumber(sourceDeviceId.serialNumber))
-                        .filter(Objects::nonNull).forEach(echoHandler -> Objects.requireNonNull(echoHandler)
-                                .handlePushActivity(currentActivity)));
+
+        /*
+         * DEPRECATED
+         * connection.getActivities(10, pushActivity.timestamp).stream().filter(activity -> search.equals(activity.id))
+         * .findFirst()
+         * .ifPresent(currentActivity -> currentActivity.getSourceDeviceIds().stream()
+         * .map(sourceDeviceId -> findEchoHandlerBySerialNumber(sourceDeviceId.serialNumber))
+         * .filter(Objects::nonNull).forEach(echoHandler -> Objects.requireNonNull(echoHandler)
+         * .handlePushActivity(currentActivity)));
+         */
+
```
remove commented code. If it is needed later, it can be retrieved from the git history.

J-N-K commented on bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java:
```diff
@@ -785,12 +786,35 @@ private void handlePushActivity(@Nullable String payload) {
+         * .handlePushActivity(currentActivity)));
+         */
+
+        Long timestamp = pushActivity.timestamp;
```
```suggestion
        long timestamp = pushActivity.timestamp;
```

use primitives where possible

Trinitus01 replied:
timestamp of pushActivity is a nullable Long so have to use Long and check for null.

J-N-K commented on bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java:
```diff
@@ -785,12 +786,35 @@ private void handlePushActivity(@Nullable String payload) {
+
+        Long timestamp = pushActivity.timestamp;
+        if (timestamp != null) {
+            Long startTimestamp = timestamp - 30000;
+            Long endTimestamp = timestamp + 30000;
```
```suggestion
            long startTimestamp = timestamp - 30000;
            long endTimestamp = timestamp + 30000;
```

J-N-K commented on bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java:
```diff
@@ -785,12 +786,35 @@ private void handlePushActivity(@Nullable String payload) {
+            for (CustomerHistoryRecord customerHistoryRecord : customerHistoryRecords) {
+                if (customerHistoryRecord != null) {
+                    String recordKey = customerHistoryRecord.recordKey;
+                    if (recordKey != null && search.equals(recordKey)) {
```
```suggestion
                    if (search.equals(recordKey)) {
```

No need to check for null: `search` is not null and .equals is always null-safe (for the second object)

Trinitus01 replied:
The check for null is not because of the following equals-method, it is because of the use of recordKey inside the if statement.

J-N-K commented on bundles/org.smarthomej.binding.amazonechocontrol/src/main/java/org/smarthomej/binding/amazonechocontrol/internal/handler/AccountHandler.java:
```diff
@@ -785,12 +786,35 @@ private void handlePushActivity(@Nullable String payload) {
+                if (customerHistoryRecord != null) {
+                    String recordKey = customerHistoryRecord.recordKey;
+                    if (recordKey != null && search.equals(recordKey)) {
+                        EchoHandler echoHandler = findEchoHandlerBySerialNumber(recordKey.split("#")[3]);
```
Is this safe? If recordKey.split returns less than four parts, this will result in an IndexOutOfBoundsException

Trinitus01 replied:
It depends on the data of the response, so i will add a check.

J-N-K reviewed (changes requested):
Thanks. Please add the new channel to README, too.

Trinitus01 committed:
added: lastSpokenText desctiption and example to readme

Signed-off-by: Tom Blum <trinitus01@googlemail.com>

Trinitus01 committed:
reverted: code structure
removed: commented code
removed: deprecated method
changed: use primitives where possible
removed: check for null
added: check for length

Signed-off-by: Tom Blum <trinitus01@googlemail.com>

Trinitus01 committed:
changed: format of readme

Signed-off-by: Tom Blum <trinitus01@googlemail.com>
